File: src/defi_sim_api/routers/registry.py
"""Registry / catalog endpoints — expose available types to the frontend."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, status

from defi_sim.engine.specs import (
    EntityMetadata,
    RunSpec,
    _CATEGORY_TABLES,
    iter_registry_categories,
)

from defi_sim_api.schemas import (
    RegistryCategoryDefinition,
    RegistryContractResponse,
    RegistryEntityDefinition,
    RunSpecSchema,
    SpecValidationResponse,
)

router = APIRouter(prefix="/registry", tags=["registry"])


# Contract version stamped on every enriched response (BE-006 tightens
# this into a contract test). Bumping this value signals to the
# frontend that a new contract shape has shipped — the frontend's
# SUPPORTED_CONTRACT_VERSION must be updated in lockstep.
REGISTRY_CONTRACT_VERSION = "v2"


# Presentation metadata used when an entity is registered without an
# EntityMetadata object. Keeps category labels stable across the
# legacy-vs-enriched transition.
_CATEGORY_LABELS: dict[str, str] = {
    "markets": "Markets",
    "agents": "Agents",
    "clocks": "Clocks",
    "orderings": "Ordering",
    "gas_models": "Cost Models",
    "fee_models": "Fee Models",
    "feeds": "Feeds",
    "execution_models": "Execution",
    "information_filters": "Information",
}

_CATEGORY_KEY_PREFIX: dict[str, str] = {
    "markets": "reg-markets",
    "agents": "reg-agents",
    "clocks": "reg-clocks",
    "orderings": "reg-ordering",
    "gas_models": "reg-gas",
    "fee_models": "reg-fees",
    "feeds": "reg-feeds",
    "execution_models": "reg-exec",
    "information_filters": "reg-information",
}


def _title_case(value: str) -> str:
    return " ".join(
        word[:1].upper() + word[1:] for word in value.replace("_", " ").split() if word
    )


def _entity_from_metadata(
    category: str, spec_type: str, meta: EntityMetadata | None
) -> RegistryEntityDefinition:
    if meta is None:
        return RegistryEntityDefinition(
            category=category,
            type=spec_type,
            label=_title_case(spec_type),
        )
    return RegistryEntityDefinition(
        category=category,
        type=spec_type,
        label=meta.label,
        description=meta.description,
        badges=list(meta.badges) if meta.badges else None,
        builder_supported=meta.builder_supported,
        schema=meta.schema,
        ui_schema=meta.ui_schema,
        defaults=meta.defaults,
        examples=list(meta.examples) if meta.examples else None,
        metadata=meta.metadata,
    )


def _build_category_definition(category: str) -> RegistryCategoryDefinition:
    tables = _CATEGORY_TABLES[category]
    factories, meta_table = tables
    entities = [
        _entity_from_metadata(category, spec_type, meta_table.get(spec_type))
        for spec_type in sorted(factories.keys())
    ]
    order = iter_registry_categories().index(category)
    return RegistryCategoryDefinition(
        key=_CATEGORY_KEY_PREFIX.get(category, f"reg-{category}"),
        label=_CATEGORY_LABELS.get(category, _title_case(category)),
        order=order,
        entities=entities,
    )


def _build_contract_response() -> RegistryContractResponse:
    return RegistryContractResponse(
        contract_version=REGISTRY_CONTRACT_VERSION,
        categories=[_build_category_definition(cat) for cat in iter_registry_categories()],
    )


@router.get(
    "",
    summary="List all registry categories and their registered types",
)
def list_all() -> dict:
    """Return the enriched v2 registry contract."""
    return _build_contract_response().model_dump(by_alias=True, exclude_none=True)


@router.get(
    "/{category}",
    summary="List registered types for a category",
)
def list_category(category: str) -> dict:
    if category not in _CATEGORY_TABLES:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Unknown registry category {category!r}. Available: {sorted(_CATEGORY_TABLES)}",
        )
    return _build_category_definition(category).model_dump(by_alias=True, exclude_none=True)


@router.post(
    "/validate",
    response_model=SpecValidationResponse,
    summary="Validate a RunSpec without executing it",
)
def validate_spec(body: RunSpecSchema) -> SpecValidationResponse:
    errors: list[str] = []
    try:
        spec = RunSpec.from_dict(body.model_dump(exclude_none=True))
    except Exception as exc:
        errors.append(str(exc))
        return SpecValidationResponse(valid=False, errors=errors)

    from defi_sim.engine.specs import build_agents, build_market, build_simulation_config

    try:
        build_market(spec.market)
        build_agents(spec.agents)
        build_simulation_config(spec)
    except Exception as exc:
        errors.append(str(exc))

    return SpecValidationResponse(valid=len(errors) == 0, errors=errors)

File: frontend/src/lib/api/adapters/registry.ts
import type { RegistryCategory, RegistryEntry } from "@/lib/types/registry";
import type { RegistryContractResponse } from "@/lib/types/contract";
import { describeType } from "@/lib/descriptions/registry";

/**
 * Backend /registry response shape (BE-002). The legacy
 * `dict[str, list[str]]` shape was removed in BE-006.
 */
export type ApiRegistryListResponse = RegistryContractResponse;

/**
 * Backend /registry/{category} response shape. Equivalent to one
 * entry in `ApiRegistryListResponse.categories`.
 */
export type ApiRegistryCategoryResponse =
  RegistryContractResponse["categories"][number];

type ApiEntity = ApiRegistryCategoryResponse["entities"][number];

type LegacyBadge = NonNullable<RegistryEntry["badges"]>[number];
type LegacyBadgeVariant = LegacyBadge["variant"];

const ALLOWED_VARIANTS: ReadonlySet<LegacyBadgeVariant> = new Set([
  "green",
  "blue",
  "purple",
  "yellow",
  "red",
]);

function coerceBadges(
  raw: ApiEntity["badges"] | null | undefined,
): RegistryEntry["badges"] {
  if (!raw) return undefined;
  return raw.map((badge) => {
    const variant = ALLOWED_VARIANTS.has(badge.variant as LegacyBadgeVariant)
      ? (badge.variant as LegacyBadgeVariant)
      : "blue";
    return { label: badge.label, variant };
  });
}

function entityToLegacyEntry(entity: ApiEntity): RegistryEntry {
  const description = entity.description ?? "";
  const base: RegistryEntry = {
    name: entity.label,
    type: entity.type,
    description: description || describeType(entity.category, entity.type).description,
  };
  const badges = coerceBadges(entity.badges);
  if (badges && badges.length > 0) base.badges = badges;
  if (entity.builderSupported === false) base.disabled = true;
  return base;
}

function categoryToLegacy(
  category: ApiRegistryCategoryResponse,
): RegistryCategory {
  return {
    key: category.key,
    label: category.label,
    description: category.description || undefined,
    entries: category.entities.map(entityToLegacyEntry),
  };
}

export function fromApiRegistry(raw: ApiRegistryListResponse): RegistryCategory[] {
  return raw.categories
    .slice()
    .sort((a, b) => (a.order ?? 0) - (b.order ?? 0))
    .map(categoryToLegacy);
}

export function fromApiRegistryCategory(
  raw: ApiRegistryCategoryResponse,
  _key: string,
): RegistryCategory {
  return categoryToLegacy(raw);
}

/** Map a frontend "reg-*" key back to the backend category string. */
const FRONTEND_TO_BACKEND: Record<string, string> = {
  "reg-markets": "markets",
  "reg-agents": "agents",
  "reg-clocks": "clocks",
  "reg-ordering": "orderings",
  "reg-gas": "gas_models",
  "reg-fees": "fee_models",
  "reg-feeds": "feeds",
  "reg-exec": "execution_models",
  "reg-information": "information_filters",
};

export function backendCategoryForKey(key: string): string {
  if (FRONTEND_TO_BACKEND[key]) return FRONTEND_TO_BACKEND[key];
  if (key.startsWith("reg-")) return key.slice("reg-".length);
  return key;
}

const BACKEND_TO_FRONTEND: Record<string, string> = Object.fromEntries(
  Object.entries(FRONTEND_TO_BACKEND).map(([k, v]) => [v, k]),
);

export function frontendKeyForBackendCategory(category: string): string {
  return BACKEND_TO_FRONTEND[category] ?? `reg-${category}`;
}
